src/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs

```diff
@@ -1098,36 +1059,30 @@ public static BigInteger ModPow(BigInteger value, BigInteger exponent, BigIntege
             exponent.AssertValid();
             modulus.AssertValid();
 
-            int signRes = +1;
-            int signVal = +1;
-            int signMod = +1;
-            bool expIsEven = exponent.IsEven;
-            BigIntegerBuilder regRes = new BigIntegerBuilder(BigInteger.One, ref signRes);
-            BigIntegerBuilder regVal = new BigIntegerBuilder(value, ref signVal);
-            BigIntegerBuilder regMod = new BigIntegerBuilder(modulus, ref signMod);
-            BigIntegerBuilder regTmp = new BigIntegerBuilder(regVal.Size);
-
-            regRes.Mod(ref regMod);   // Handle special case of exponent=0, modulus=1
+            bool trivialValue = value._bits == null;
+            bool trivialExponent = exponent._bits == null;
+            bool trivialModulus = modulus._bits == null;
 
-            if (exponent._bits == null)
-            {   // exponent fits into an Int32
-                ModPowInner((uint)exponent._sign, ref regRes, ref regVal, ref regMod, ref regTmp);
-            }
-            else
-            {   // very large exponent
-                int len = Length(exponent._bits);
-                for (int i = 0; i < len - 1; i++)
-                {
-                    uint exp = exponent._bits[i];
-                    ModPowInner32(exp, ref regRes, ref regVal, ref regMod, ref regTmp);
-                }
-                ModPowInner(exponent._bits[len - 1], ref regRes, ref regVal, ref regMod, ref regTmp);
-            }
+            uint[] bits = trivialValue && trivialExponent && trivialModulus
+                        ? BigIntegerCalculator.Pow(NumericsHelpers.Abs(value._sign), NumericsHelpers.Abs(exponent._sign), NumericsHelpers.Abs(modulus._sign))
+                        : trivialExponent && trivialModulus
+                        ? BigIntegerCalculator.Pow(value._bits, NumericsHelpers.Abs(exponent._sign), NumericsHelpers.Abs(modulus._sign))
+                        : trivialValue && trivialModulus
+                        ? BigIntegerCalculator.Pow(NumericsHelpers.Abs(value._sign), exponent._bits, NumericsHelpers.Abs(modulus._sign))
+                        : trivialValue && trivialExponent
+                        ? BigIntegerCalculator.Pow(NumericsHelpers.Abs(value._sign), NumericsHelpers.Abs(exponent._sign), modulus._bits)
+                        : trivialValue
+                        ? BigIntegerCalculator.Pow(NumericsHelpers.Abs(value._sign), exponent._bits, modulus._bits)
+                        : trivialExponent
+                        ? BigIntegerCalculator.Pow(value._bits, NumericsHelpers.Abs(exponent._sign), modulus._bits)
+                        : trivialModulus
+                        ? BigIntegerCalculator.Pow(value._bits, exponent._bits, NumericsHelpers.Abs(modulus._sign))
+                        : BigIntegerCalculator.Pow(value._bits, exponent._bits, modulus._bits);
```

Comment:
That said, I wonder if in this particular case it would be both slightly better performing and easier to read if it were instead broken up into a few pieces like this (you'll want to double-check that I haven't messed up the logic here in copying-and-pasting):
```C#
uint[] bits;
if (trivialValue)
{
    bits = 
        trivialExponent & trivialModulus ? BigIntegerCalculator.Pow(NumericsHelpers.Abs(value._sign), NumericsHelpers.Abs(exponent._sign), NumericsHelpers.Abs(modulus._sign)) :
        trivialModulus ? BigIntegerCalculator.Pow(NumericsHelpers.Abs(value._sign), exponent._bits, NumericsHelpers.Abs(modulus._sign)) :
        trivialExponent ? BigIntegerCalculator.Pow(NumericsHelpers.Abs(value._sign), NumericsHelpers.Abs(exponent._sign), modulus._bits) :
        BigIntegerCalculator.Pow(NumericsHelpers.Abs(value._sign), exponent._bits, modulus._bits);
}
else if (trivialExponent)
{
    bits = trivialModulus ?
        BigIntegerCalculator.Pow(value._bits, NumericsHelpers.Abs(exponent._sign), NumericsHelpers.Abs(modulus._sign)) :
        BigIntegerCalculator.Pow(value._bits, NumericsHelpers.Abs(exponent._sign), modulus._bits);
}
else
{
    bits = trivialModulus ?
        BigIntegerCalculator.Pow(value._bits, exponent._bits, NumericsHelpers.Abs(modulus._sign)) :
        BigIntegerCalculator.Pow(value._bits, exponent._bits, modulus._bits);
}
```